### PR TITLE
Implement io.edgehog.devicemanager.config.Telemetry

### LINKED ISSUE #11: Implement `io.edgehog.devicemanager.config.Telemetry` Interface in Device Runtime
See also: https://github.com/edgehog-device-manager/edgehog-astarte-interfaces

### PR DESCRIPTION
**Pull request by FrancescoVaiani**
- Add support to `io.edgehog.devicemanager.config.Telemetry`
- Add `telemetry_config` as an array of mappings in the `edgehog_config.toml` to represent the default telemetry configs.

Close #11

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -32,3 +32,4 @@ uuid = {version="0.8.2", features = ["v5", "v4"] }
 systemd = { version = "0.10", optional = true }
 mockall = "0.11.0"
 async-trait = "0.1.53"
+mockall_double = "0.3.0"
```
Comment (harlem88) [inline Cargo.toml:35]: Remove this if it's no longer used.

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/data/astarte.rs [hunk 2]
```diff
@@ -32,7 +32,7 @@ pub struct Astarte {
 
 #[async_trait]
 impl Publisher for Astarte {
-    async fn send_object<T>(
+    async fn send_object<T: 'static>(
```
Comment (rbino) [inline src/data/astarte.rs:35]: If I understand correctly, isn't this saying that _all_ payloads that we pass to `send_object` have a `static` lifetime (so they're valid throughout the course of all the program)? Is this actually the case?

File: src/telemetry/mod.rs [hunk 3]
```diff
@@ -18,7 +18,413 @@
...
+                self.unset_enabled(interface_name).await;
+            }
+
+            ("periodSeconds", AstarteType::LongInteger(period)) => {
+                self.set_period(interface_name, *period as u64).await;
+            }
+
+            ("periodSeconds", AstarteType::Integer(period)) => {
+                self.set_period(interface_name, *period as u64).await;
+            }
```
Comment (rbino) [inline src/telemetry/mod.rs:245]: Why are you receiving both integer and longinteger for the same endpoint? Shouldn't it be either one of them?

**Review by rbino — CHANGES_REQUESTED** [review]

File: src/data/astarte.rs (hunk 2 above)
Comment (harlem88) [reply src/data/astarte.rs:35]: @rbino It related to this case [trait-bound](https://doc.rust-lang.org/rust-by-example/scope/lifetime/static_lifetime.html#trait-bound)

File: src/lib.rs [hunk 4]
```diff
@@ -18,21 +18,23 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+use crate::data::astarte::Astarte;
+use crate::data::Publisher;
+use crate::ota::ota_handler::OTAHandler;
+use crate::telemetry::TelemetryPayload;
```
Comment (harlem88) [inline src/lib.rs:24]: Why did you move here this import? IMHO the imports should be split into at least three groups, consisting of:
- imports from `std`
- imports from external crates
- imports from within the same crate

Refer to [StdExternalCrate](https://rust-lang.github.io/rustfmt/?version=v1.4.30&search=group#StdExternalCrate%5C%3A)
what do you think @bettio @rbino ?

File: src/telemetry/mod.rs [hunk 5]
```diff
@@ -18,8 +18,414 @@
...
+
+pub struct Telemetry {
+    telemetry_task_configs: Arc<RwLock<HashMap<String, TelemetryTaskConfig>>>,
+    kill_switches: HashMap<String, Sender<()>>,
+    communication_channel: MpscSender<TelemetryPayload>,
+}
+
+pub enum TelemetryPayload {
+    SystemStatus(crate::telemetry::system_status::SystemStatus),
+}
```
Comment (harlem88) [inline src/telemetry/mod.rs:65]: Add `StorageUsage` to the telemetry available.

File: src/telemetry/mod.rs [hunk 6]
```diff
@@ -18,8 +18,414 @@
...
+
+async fn send_data(communication_channel: &MpscSender<TelemetryPayload>, interface_name: String) {
+    debug!("sending {interface_name}");
+
+    if interface_name.as_str() == "io.edgehog.devicemanager.SystemStatus" {
+        let sysstatus = system_status::get_system_status().unwrap();
+        let _ = communication_channel
+            .send(TelemetryPayload::SystemStatus(sysstatus))
+            .await;
+    }
```
Comment (harlem88) [inline src/telemetry/mod.rs:287]: Add also here `StorageUsage` telemetry.

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/telemetry/mod.rs [hunk 7]
```diff
@@ -18,8 +18,437 @@
...
+                    default_enabled: c.enabled.clone(),
+                    default_period: c.period.clone(),
+                    override_enabled: None,
+                    override_period: None,
+                },
+            );
+        }
+
+        // Serde load configs from disk
+        if let Ok(file) = File::open(TELEMETRY_PATH) {
```
Comment (harlem88) [inline src/telemetry/mod.rs:92]: Use StateRepository to persist TelemetryInterfaceConfig

File: src/telemetry/mod.rs [hunk 8]
```diff
@@ -18,8 +18,437 @@
...
+        {
+            let interface_config = TelemetryInterfaceConfig {
+                interface_name: interface_name.to_string(),
+                enabled: telemetry_task_config.override_enabled,
+                period: telemetry_task_config.override_period,
+            };
+
+            telemetry_config.push(interface_config);
+        }
+        let file = File::create(TELEMETRY_PATH).unwrap();
```
Comment (harlem88) [inline src/telemetry/mod.rs:280]: Same above

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by rbino — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Review by bettio — APPROVED** [review]
